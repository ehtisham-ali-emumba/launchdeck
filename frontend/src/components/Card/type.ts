export interface CardProps {
  imageSrc: string;
  onClick?: () => void;
  imageHeight: number;
  dimensions: {
    width: number;
    height: number;
  };
  imageStyles?: React.CSSProperties;
  footerStyles?: React.CSSProperties;
  renderLowerLeft?: () => React.ReactNode;
  renderLowerRight?: () => React.ReactNode;
  renderUpperLeft?: () => React.ReactNode;
  renderUpperRight?: () => React.ReactNode;
  renderContent?: () => React.ReactNode;
  url: string;
}

export type StyledBaseCardType = {
  dimensions: {
    width: number;
    height: number;
  };
  imageHeight: number;
};
