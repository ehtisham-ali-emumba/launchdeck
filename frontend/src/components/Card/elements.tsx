import { Card } from "antd";
import styled, { css } from "styled-components";

import type { StyledBaseCardType } from "./type";

export const StyledBaseCard = styled(Card)<StyledBaseCardType>`
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.06);
  height: ${({ dimensions }) => dimensions.height}px;
  width: ${({ dimensions }) => dimensions.width}px;
  margin: 0 auto;
  position: relative;

  .ant-card-cover img {
    height: ${({ imageHeight }) => imageHeight}px;
  }
  .ant-card-body {
    padding: 8px;
  }
`;

export const BaseCardImageWrapper = styled.div<{ imageHeight: number }>`
  position: relative;
  min-height: ${({ imageHeight }) => imageHeight}px;
  height: ${({ imageHeight }) => imageHeight}px;
  align-items: center;
  flex: 1;
  justify-content: center;
  overflow: hidden;
  && {
    display: flex;
  }
`;
export const BaseCardImage = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
  transition: transform 0.3s ease;
  ${StyledBaseCard}:hover & {
    transform: scale(1.05);
  }
`;

export const BaseCardContent = styled.div``;

const upperText = css`
  position: absolute;
  top: 8px;
  transition: inherit;
`;
export const BaseCardUpperLeft = styled.div`
  ${upperText} left: 10px;
`;
export const BaseCardUpperRight = styled.div`
  ${upperText} right: 10px;
`;

export const FooterWrapper = styled.div`
  position: absolute;
  bottom: 8px;
  left: 0;
  right: 0;
  padding: 0 16px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;
