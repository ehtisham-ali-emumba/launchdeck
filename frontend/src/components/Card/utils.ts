export const cardDimensions = {
  width: 300,
  height: 415,
};
