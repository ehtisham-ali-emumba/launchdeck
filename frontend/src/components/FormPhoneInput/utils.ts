export const countryCodeArray = [
  {
    value: "+1",
    label: "+1 (USA)",
  },
  {
    value: "+86",
    label: "+86 (China)",
  },
  {
    value: "+92",
    label: "+92 (Pak)",
  },
];
