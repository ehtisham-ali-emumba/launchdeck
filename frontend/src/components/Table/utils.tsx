export const createPagination = (pageSize = 10, showSizeChanger = false) => {
  return {
    pageSize,
    showSizeChanger,
    showTotal: (total: number, range: [number, number]) =>
      `${range[0]}-${range[1]} of ${total} items`,
  };
};

export const getDateSorter = (dateField: string) => {
  return (a: Record<string, any>, b: Record<string, any>) => {
    const dateA = a[dateField] ? new Date(a[dateField]).getTime() : 0;
    const dateB = b[dateField] ? new Date(b[dateField]).getTime() : 0;
    return dateA - dateB;
  };
};

export const getNumberSorter = (numField: string) => {
  return (a: Record<string, any>, b: Record<string, any>) => {
    const numA = a[numField] || 0;
    const numB = b[numField] || 0;
    return numA - numB;
  };
};

export const getStringSorter = (strField: string) => {
  return (a: Record<string, any>, b: Record<string, any>) => {
    const strA = a[strField] || "";
    const strB = b[strField] || "";
    return strA.localeCompare(strB);
  };
};

export const formatDate = (
  date: string,
  options: Intl.DateTimeFormatOptions = {
    year: "numeric",
    month: "short",
    day: "numeric",
  }
) => {
  if (!date) return "";
  return new Date(date).toLocaleDateString(undefined, options);
};

export const renderNumericValue = (value: number, highlight = false) => (
  <span style={{ fontWeight: highlight ? "600" : "normal" }}>{value}</span>
);

export const getDefaultRowKey = (record: any, index?: number) =>
  record.id || record._id || index;
