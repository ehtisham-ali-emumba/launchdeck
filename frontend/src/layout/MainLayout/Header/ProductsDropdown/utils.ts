import type { Category } from "./types";

export const categoriesData: Category[] = [
  {
    id: "engineering",
    name: "Engineering & Development",
    description: "Tools for software development",
    subCategories: [
      {
        id: "frontend",
        name: "Frontend Development",
        description: "UI/UX tools and frameworks",
        slug: "frontend-development",
        icon: "🎨",
      },
      {
        id: "backend",
        name: "Backend Development",
        description: "Server-side tools and APIs",
        slug: "backend-development",
        icon: "⚙️",
      },
      {
        id: "mobile",
        name: "Mobile Development",
        description: "iOS and Android development",
        slug: "mobile-development",
        icon: "📱",
      },
      {
        id: "devops",
        name: "DevOps & Infrastructure",
        description: "Deployment and monitoring tools",
        slug: "devops-infrastructure",
        icon: "🚀",
      },
    ],
  },
  {
    id: "design",
    name: "Design & Creative",
    description: "Creative tools and resources",
    subCategories: [
      {
        id: "ui-design",
        name: "UI/UX Design",
        description: "Interface design tools",
        slug: "ui-ux-design",
        icon: "🎯",
      },
      {
        id: "graphic-design",
        name: "Graphic Design",
        description: "Visual design tools",
        slug: "graphic-design",
        icon: "🖼️",
      },
      {
        id: "prototyping",
        name: "Prototyping",
        description: "Rapid prototyping tools",
        slug: "prototyping",
        icon: "🔧",
      },
    ],
  },
  {
    id: "productivity",
    name: "Productivity & Business",
    description: "Tools to boost productivity",
    subCategories: [
      {
        id: "project-management",
        name: "Project Management",
        description: "Organize and track projects",
        slug: "project-management",
        icon: "📊",
      },
      {
        id: "collaboration",
        name: "Team Collaboration",
        description: "Communication and teamwork",
        slug: "team-collaboration",
        icon: "👥",
      },
      {
        id: "analytics",
        name: "Analytics & Data",
        description: "Data analysis and insights",
        slug: "analytics-data",
        icon: "📈",
      },
    ],
  },
  {
    id: "marketing",
    name: "Marketing & Sales",
    description: "Grow your business",
    subCategories: [
      {
        id: "social-media",
        name: "Social Media",
        description: "Social media management",
        slug: "social-media",
        icon: "📢",
      },
      {
        id: "email-marketing",
        name: "Email Marketing",
        description: "Email campaigns and automation",
        slug: "email-marketing",
        icon: "📧",
      },
      {
        id: "seo",
        name: "SEO & Content",
        description: "Search optimization tools",
        slug: "seo-content",
        icon: "🔍",
      },
    ],
  },
  {
    id: "engineering",
    name: "Engineering & Development",
    description: "Tools for software development",
    subCategories: [
      {
        id: "frontend",
        name: "Frontend Development",
        description: "UI/UX tools and frameworks",
        slug: "frontend-development",
        icon: "🎨",
      },
      {
        id: "backend",
        name: "Backend Development",
        description: "Server-side tools and APIs",
        slug: "backend-development",
        icon: "⚙️",
      },
      {
        id: "mobile",
        name: "Mobile Development",
        description: "iOS and Android development",
        slug: "mobile-development",
        icon: "📱",
      },
      {
        id: "devops",
        name: "DevOps & Infrastructure",
        description: "Deployment and monitoring tools",
        slug: "devops-infrastructure",
        icon: "🚀",
      },
    ],
  },
  {
    id: "design",
    name: "Design & Creative",
    description: "Creative tools and resources",
    subCategories: [
      {
        id: "ui-design",
        name: "UI/UX Design",
        description: "Interface design tools",
        slug: "ui-ux-design",
        icon: "🎯",
      },
      {
        id: "graphic-design",
        name: "Graphic Design",
        description: "Visual design tools",
        slug: "graphic-design",
        icon: "🖼️",
      },
      {
        id: "prototyping",
        name: "Prototyping",
        description: "Rapid prototyping tools",
        slug: "prototyping",
        icon: "🔧",
      },
    ],
  },
];
