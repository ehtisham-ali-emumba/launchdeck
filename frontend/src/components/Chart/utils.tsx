import type { ChartData, ChartOptions } from "chart.js";

import type { TSFixMe } from "~/types";

import { CHART_COLORS } from "./elements";

const generateColors = (count: number, alpha?: number): string[] => {
  const colors = CHART_COLORS.defaultColors;
  const result: string[] = [];

  // Loop through the colors and repeat if needed
  for (let i = 0; i < count; i++) {
    const baseColor = colors[i % colors.length];

    if (alpha !== undefined) {
      // Add alpha channel if specified
      const rgbMatch = baseColor.match(
        /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i
      );

      if (rgbMatch) {
        const r = parseInt(rgbMatch[1], 16);
        const g = parseInt(rgbMatch[2], 16);
        const b = parseInt(rgbMatch[3], 16);
        result.push(`rgba(${r}, ${g}, ${b}, ${alpha})`);
      } else {
        result.push(baseColor);
      }
    } else {
      result.push(baseColor);
    }
  }

  return result;
};

export const createBarChartData = (
  labels: string[],
  data: number[],
  label = "Value"
): ChartData<"bar"> => {
  return {
    labels,
    datasets: [
      {
        label,
        data,
        backgroundColor: generateColors(data.length),
        borderColor: generateColors(data.length, 0.8),
        borderWidth: 1,
        borderRadius: 4,
      },
    ],
  };
};

export const createPieChartData = (
  labels: string[],
  data: number[]
): ChartData<"pie"> => {
  return {
    labels,
    datasets: [
      {
        data,
        backgroundColor: generateColors(data.length),
        borderColor: "#fff",
        borderWidth: 2,
      },
    ],
  };
};

export const createLineChartData = (
  labels: string[],
  data: number[],
  label = "Value"
): ChartData<"line"> => {
  return {
    labels,
    datasets: [
      {
        label,
        data,
        borderColor: CHART_COLORS.line.border,
        backgroundColor: CHART_COLORS.line.background,
        fill: true,
        tension: 0.4,
        pointBackgroundColor: CHART_COLORS.line.point,
        pointBorderColor: CHART_COLORS.line.pointBorder,
        pointBorderWidth: 2,
        pointRadius: 4,
      },
    ],
  };
};

export const defaultBarOptions: ChartOptions<"bar"> = {
  responsive: true,
  maintainAspectRatio: false,
  plugins: {
    legend: {
      position: "top",
      align: "end",
      labels: {
        boxWidth: 15,
        usePointStyle: true,
        padding: 15,
      },
    },
    tooltip: {
      backgroundColor: "rgba(0, 0, 0, 0.8)",
      titleColor: "#fff",
      bodyColor: "#fff",
      padding: 12,
      cornerRadius: 4,
    },
  },
  scales: {
    y: {
      beginAtZero: true,
      grid: {
        color: "rgba(0, 0, 0, 0.1)",
      },
    },
    x: {
      grid: {
        display: false,
      },
    },
  },
};

export const defaultLineOptions: ChartOptions<"line"> = {
  responsive: true,
  maintainAspectRatio: false,
  plugins: {
    legend: {
      position: "top",
      align: "end",
      labels: {
        boxWidth: 15,
        usePointStyle: true,
        padding: 15,
      },
    },
    tooltip: {
      backgroundColor: "rgba(0, 0, 0, 0.8)",
      titleColor: "#fff",
      bodyColor: "#fff",
      padding: 12,
      cornerRadius: 4,
    },
  },
  scales: {
    y: {
      beginAtZero: true,
      grid: {
        color: "rgba(0, 0, 0, 0.1)",
      },
    },
    x: {
      grid: {
        color: "rgba(0, 0, 0, 0.05)",
      },
    },
  },
};

export const defaultPieOptions: ChartOptions<"pie"> = {
  responsive: true,
  maintainAspectRatio: false,
  plugins: {
    legend: {
      position: "right",
      align: "center",
      labels: {
        boxWidth: 15,
        usePointStyle: true,
        padding: 15,
      },
    },
    tooltip: {
      backgroundColor: "rgba(0, 0, 0, 0.8)",
      titleColor: "#fff",
      bodyColor: "#fff",
      padding: 12,
      cornerRadius: 4,
    },
  },
};

const defaultDoughnutOptions: ChartOptions<"doughnut"> = {
  ...(defaultPieOptions as ChartOptions<"doughnut">),
  cutout: "70%",
};

const defaultRadarOptions: ChartOptions<"radar"> = {
  responsive: true,
  maintainAspectRatio: false,
  elements: {
    line: {
      tension: 0.2,
    },
    point: {
      radius: 4,
    },
  },
  plugins: {
    legend: {
      position: "top",
      labels: {
        boxWidth: 15,
        usePointStyle: true,
        padding: 15,
      },
    },
    tooltip: {
      backgroundColor: "rgba(0, 0, 0, 0.8)",
      titleColor: "#fff",
      bodyColor: "#fff",
      padding: 12,
      cornerRadius: 4,
    },
  },
  scales: {
    r: {
      beginAtZero: true,
      angleLines: {
        color: "rgba(0, 0, 0, 0.1)",
      },
      grid: {
        color: "rgba(0, 0, 0, 0.1)",
      },
      pointLabels: {
        font: {
          size: 12,
        },
      },
    },
  },
};

const getDefaultOptions = (type: string): ChartOptions<TSFixMe> => {
  switch (type) {
    case "bar":
      return defaultBarOptions;
    case "line":
      return defaultLineOptions;
    case "pie":
      return defaultPieOptions;
    case "doughnut":
      return defaultDoughnutOptions;
    case "radar":
      return defaultRadarOptions;
    default:
      return defaultBarOptions;
  }
};

export const mergeChartOptions = (
  type: string,
  options?: ChartOptions<TSFixMe>
): ChartOptions<TSFixMe> => {
  const defaultOptions = getDefaultOptions(type);
  return {
    ...defaultOptions,
    ...(options || {}),
    plugins: {
      ...defaultOptions.plugins,
      ...(options?.plugins || {}),
    },
    scales: {
      ...defaultOptions.scales,
      ...(options?.scales || {}),
    },
  };
};
