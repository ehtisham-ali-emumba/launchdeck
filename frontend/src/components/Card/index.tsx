import React from "react";

import { BaseLink } from "../elements";

import {
  StyledBaseCard,
  BaseCardImageWrapper,
  BaseCardImage,
  BaseCardContent,
  BaseCardUpperLeft,
  BaseCardUpperRight,
  FooterWrapper,
} from "./elements";
import type { CardProps } from "./type";
import { cardDimensions } from "./utils";

export const Card: React.FC<CardProps> = ({
  imageSrc,
  onClick,
  imageHeight = 200,
  dimensions = cardDimensions,
  imageStyles = {},
  footerStyles = {},
  renderLowerLeft = () => null,
  renderLowerRight = () => null,
  renderUpperLeft = () => null,
  renderUpperRight = () => null,
  renderContent = () => null,
  url,
}) => {
  return (
    <BaseLink to={url} onClick={onClick}>
      <StyledBaseCard
        dimensions={dimensions}
        imageHeight={imageHeight}
        hoverable={false}
        cover={
          <BaseCardImageWrapper imageHeight={imageHeight}>
            <BaseCardImage src={imageSrc} style={imageStyles} />
          </BaseCardImageWrapper>
        }
      >
        <BaseCardContent>{renderContent()}</BaseCardContent>
        <BaseCardUpperLeft>{renderUpperLeft()}</BaseCardUpperLeft>
        <BaseCardUpperRight>{renderUpperRight()}</BaseCardUpperRight>
        <FooterWrapper style={footerStyles}>
          {renderLowerLeft()}
          {renderLowerRight()}
        </FooterWrapper>
      </StyledBaseCard>
    </BaseLink>
  );
};
