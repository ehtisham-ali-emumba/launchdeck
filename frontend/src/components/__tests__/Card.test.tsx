import { MemoryRouter } from "react-router-dom";

import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";

import { Card } from "../Card";
import { cardDimensions } from "../Card/utils";

const defaultProps = {
  imageSrc: "test-image.jpg",
  url: "/test-url",
  imageHeight: 200,
  dimensions: cardDimensions,
};

describe("Card", () => {
  const renderWithRouter = (ui: React.ReactElement) =>
    render(<MemoryRouter>{ui}</MemoryRouter>);

  it("matches snapshot", () => {
    const { container } = renderWithRouter(<Card {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders image and navigates to url", () => {
    renderWithRouter(<Card {...defaultProps} />);
    const img = screen.getByRole("img");
    expect(img).toHaveAttribute("src", defaultProps.imageSrc);
    const link = screen.getByRole("link");
    expect(link).toHaveAttribute("href", defaultProps.url);
  });

  it("calls onClick when clicked", async () => {
    const user = userEvent.setup();
    const handleClick = vi.fn();
    renderWithRouter(<Card {...defaultProps} onClick={handleClick} />);
    const link = screen.getByRole("link");
    await user.click(link);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("renders custom content and footer", () => {
    renderWithRouter(
      <Card
        {...defaultProps}
        renderContent={() => <div>Card Content</div>}
        renderLowerLeft={() => <span>Left</span>}
        renderLowerRight={() => <span>Right</span>}
        renderUpperLeft={() => <span>ULeft</span>}
        renderUpperRight={() => <span>URight</span>}
      />
    );
    expect(screen.getByText("Card Content")).toBeInTheDocument();
    expect(screen.getByText("Left")).toBeInTheDocument();
    expect(screen.getByText("Right")).toBeInTheDocument();
    expect(screen.getByText("ULeft")).toBeInTheDocument();
    expect(screen.getByText("URight")).toBeInTheDocument();
  });

  it("applies correct dimensions to the card", () => {
    renderWithRouter(<Card {...defaultProps} />);
    const card = screen.getByRole("img").closest(".ant-card");
    expect(card).toHaveStyle(`width: ${defaultProps.dimensions.width}px`);
    expect(card).toHaveStyle(`height: ${defaultProps.dimensions.height}px`);
  });
});
