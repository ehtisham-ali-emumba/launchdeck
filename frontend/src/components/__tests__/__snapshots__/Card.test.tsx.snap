// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card > matches snapshot 1`] = `
<div>
  <a
    class="sc-fYsHOw jQoQIn"
    data-discover="true"
    href="/test-url"
  >
    <div
      class="ant-card ant-card-bordered sc-fWnslK kPhQvw css-dev-only-do-not-override-1m63z2v"
      dimensions="[object Object]"
      imageheight="200"
    >
      <div
        class="ant-card-cover"
      >
        <div
          class="sc-dIMoHT btnOzl"
          imageheight="200"
        >
          <img
            class="sc-iQQCXo cnOgwE"
            src="test-image.jpg"
          />
        </div>
      </div>
      <div
        class="ant-card-body"
      >
        <div
          class="sc-gDpztx"
        />
        <div
          class="sc-kpOvIu huDXwc"
        />
        <div
          class="sc-icnseD hPcLor"
        />
        <div
          class="sc-jMsorb foHhwn"
        />
      </div>
    </div>
  </a>
</div>
`;
